deploy-cloud-functions/src/cloudFunctionClient.ts

```diff
@@ -0,0 +1,337 @@
+    const getRequest: cloudfunctions_v1.Params$Resource$Projects$Locations$Functions$Get = {
+      name: fqn,
+      auth: authClient,
+    };
+    const currentFunctionResponse: GaxiosResponse<cloudfunctions_v1.Schema$CloudFunction> = await this.gcf.projects.locations.functions.get(
+      getRequest,
+      this.methodOptions,
+    );
+    return currentFunctionResponse.data;
+  }
+
+  /**
+   * Retrieves a list of currently deployed Cloud Functions.
+   *
+   * @returns list of Cloud Functions.
+   */
+  async getFunctions(): Promise<string[]> {
+    const authClient = await this.getAuthClient();
+    const getRequest: cloudfunctions_v1.Params$Resource$Projects$Locations$Functions$List = {
+      parent: this.parent,
+      auth: authClient,
+    };
+    const currentFunctionsResponse: GaxiosResponse<cloudfunctions_v1.Schema$ListFunctionsResponse> = await this.gcf.projects.locations.functions.list(
+      getRequest,
+      this.methodOptions,
+    );
+    const currentFunctions: cloudfunctions_v1.Schema$ListFunctionsResponse =
+      currentFunctionsResponse.data;
+    let currentFunctionsNames: string[] = [];
+    if (currentFunctions.functions?.length !== undefined) {
+      currentFunctionsNames = currentFunctions.functions?.map(
+        (currentFunction) => currentFunction.name as string,
+      );
+    }
+    return currentFunctionsNames;
+  }
+
+  /**
+   * Returns a signed URL to upload source code.
+   *
+   * @returns signed url.
+   */
+  async getUploadUrl(): Promise<
+    cloudfunctions_v1.Schema$GenerateUploadUrlResponse
+  > {
+    const authClient = await this.getAuthClient();
+    const getRequest: cloudfunctions_v1.Params$Resource$Projects$Locations$Functions$Generateuploadurl = {
+      parent: this.parent,
+      auth: authClient,
+    };
+    const uploadUrlResponse: GaxiosResponse<cloudfunctions_v1.Schema$GenerateUploadUrlResponse> = await this.gcf.projects.locations.functions.generateUploadUrl(
+      getRequest,
+      this.methodOptions,
+    );
+    return uploadUrlResponse.data;
+  }
+
+  /**
+   * Deploy a CloudFunction.
+   *
+   * @param cf Cloud Function.
+   * @returns CloudFunction deployment response.
+   */
+  async deploy(cf: CloudFunction): Promise<cloudfunctions_v1.Schema$Operation> {
+    const authClient = await this.getAuthClient();
+    const deployedFunctions = await this.getFunctions();
+    const zipPath = await zipDir(cf.sourceDir);
+    const uploadUrl = await this.getUploadUrl();
+    if (!uploadUrl.uploadUrl) {
+      throw new Error('Unable to generate signed Url');
+    }
+    // Upload source code
+    await uploadSource(uploadUrl.uploadUrl, zipPath);
+    // Delete temp zip file after upload
+    await deleteZipFile(zipPath);
+    cf.setSourceUrl(uploadUrl.uploadUrl);
+    // If CF already exists, create a revision else create new deployment
+    if (deployedFunctions.includes(cf.fqn)) {
+      core.info('Creating a function revision');
+      const updateMasks = [
```

Comment:
Added a comment. So IIUC this is used so that we can update individual fields for CF, for instance just updating envvars will not cause a redeploy. In our case I assume that this action will be triggered on a commit that changes some files (which users can control with `push.path`) and hence I redeploy it. Since the whole function is defined as code, we don't need to selectively mask. Happy to discuss this further if you have concerns.